### ISSUE TITLE
Swift <-> Obj-C Bridging Conflict

### ISSUE THREAD

**Issue by Tornquist** [twilio/twilio-video-ios#109]
### Description

When building an Xcode project with both Swift and Obj-C, and with interfaces exposed to each other, TwilioVideo creates deadlocks when including. This can be resolved through additional imports, but generates persistent warnings.

### Steps to Reproduce

1. Create swift project
2. Create bridging header to expose Obj-C interface to swift
3. Create obj-c generated interface header to expose Swift to Obj-C
4. Import filename from step 3 in prefix header (actually exposes interface)
5. Create forward declarations header to expose specific classes to Obj-C prior to swift compilation
6. Install TwilioSDK (following any of the supported methods). Problem applies to Carthage, CocoaPods, manual (dynamic) and manual (static)
7. Build project

Error: `Cannot find protocol declaration for TVIVideoViewDelegate` in `MyProject-Swift.h` (the generated interface header). This error makes general sense, but is problematic. In the build process above the following will occur.

1. Grab some swift class names from the forward declarations
2. Build Obj-C project
3. Expose Obj-C back to Swift using bridging header
4. Expose Swift implementation back to Obj-C with the prefix header.

The renaming/aliasing of TVIVideoViewDelegate in Obj-C vs VideoViewDelegate in Swift breaks with this dependency map. It's unable to find the required files to actually build.

This can be solved based on importing `TwilioVideo.h` in the prefix header which allows everything to build and function as-expected but results in the following warnings:

```
Duplicate definition of category 'AVCaptureDeviceControl' on interface 'TVICameraSource'
Duplicate definition of category 'CallKit' on interface 'TVIConnectOptionsBuilder'
Duplicate definition of category 'CallKit' on interface 'TVIConnectOptions'
Duplicate definition of category 'CallKit' on interface 'TVIRoom'
```

All of the above have an import in the original header file (which is grabbed through `TwilioVideo.h` and in the `@import` of TwilioVideo

#### Code

_PrefixHeader.pch_
```c
#ifndef MyProject_PrefixHeader_pch
#define MyProject_PrefixHeader_pch

#import <MapKit/MapKit.h>
#import <CoreData/CoreData.h>
#import <Stripe/Stripe.h>
#import <Charts/Charts.h>
#import <MessageUI/MessageUI.h>
#import <AVFoundation/AVFoundation.h>
#import "TwilioVideo.h"
#import "MyProject-Swift.h"

#endif
```

_MyProject-Bridging-Header.h_
```c
... a small handful of 100% Obj-C header imports to expose to swift
```

_MyProject-Swift-ForwardDeclarations.h_
```
@class MyClass1;
@class MyClass2;
```

_MyProject-Swift.h_
```
Auto-generated
```

#### Expected Behavior

Some method to use TwilioVideo SDK in projects that already use both Obj-C and Swift (and have communication channels between both sides of the stack) without generating additional warnings.

#### Actual Behavior

Additional warnings are generated (due to duplicate header imports)

#### Reproduces How Often

100%

#### Logs

Sample of paths and warning shared from one of the warnings generated:

* /MyProject/Dependencies/TwilioVideo.framework/Headers/TVICameraSource.h:261:12: Duplicate definition of category 'AVCaptureDeviceControl' on interface 'TVICameraSource'
    * /Users/nathan/Library/Developer/Xcode/DerivedData/MyProject-asuagprksufwcsgbhzpyqtcmczux/Build/Intermediates.noindex/MyProject.build/Debug-iphoneos/MyProject.build/DerivedSources/MyProject-Swift.h:197:2: In module 'TwilioVideo' imported from /Users/nathan/Library/Developer/Xcode/DerivedData/MyProject-asuagprksufwcsgbhzpyqtcmczux/Build/Intermediates.noindex/MyProject.build/Debug-iphoneos/MyProject.build/DerivedSources/MyProject-Swift.h:197:
    * /MyProject/Dependencies/TwilioVideo.framework/Headers/TVICameraSource.h:261:12: Previous definition is here

#### Video iOS SDK

Tested with 3.2.4

#### Xcode

Using 11.5 now. Impacted 11.4 as well.

#### iOS Version

All -- Build issue

#### iOS Device

All -- Build issue

**Comment by timrozum** [comment]
Hi @Tornquist,

I think some lines in PrefixHeader.pch may be contributing to the problem. 

I would recommend moving MyProject-Swift.h out of PrefixHeader.pch and only import it in the files that really need it. I think this is conventional for the Swift header and it seems like adding the Swift header to all files (via PrefixHeader.pch) could cause problems. 

I would also do the same with TwilioVideo.h until the errors and warnings are resolved.

I don't fully understand the intent of MyProject-Swift-ForwardDeclarations.h but again it may be less problematic to only add forward declarations exactly where they are needed. 

Other than that it seems pretty standard. 

To summarize I would recommend particularly when it comes to Swift and Objective-C bridging to be as surgical and precise as possible with imports because bridging adds another layer of complexity where things can go wrong.

We also have an example video app with a little bridging that remains which may be helpful: https://github.com/twilio/twilio-video-app-ios

Let me know if this helps!

**Comment by timrozum** [comment]
I'm going to close this out for now @Tornquist but let me know if you have any other questions.

**Comment by Tornquist** [comment]
@timrozum I'm sorry for not getting back to you. Thank you for your response!

I passed it on to the team that'll be fixing this and forgot to respond back here. I do not believe that we have other questions at this time.

Thanks again!

**Comment by timrozum** [comment]
No problem @Tornquist and thanks for the update! We're here if anything else comes up.
